src/cascadia/QueryExtension/GithubCopilotLLMProvider.cpp

```diff
@@ -0,0 +1,349 @@
+static constexpr std::wstring_view apiKey{ L"api" };
+static constexpr std::wstring_view viewerKey{ L"viewer" };
+static constexpr std::wstring_view copilotEndpointsKey{ L"copilotEndpoints" };
+static constexpr std::wstring_view loginKey{ L"login" };
+static constexpr std::wstring_view grantTypeKey{ L"grant_type" };
+static constexpr std::wstring_view contentKey{ L"content" };
+static constexpr std::wstring_view messageKey{ L"message" };
+static constexpr std::wstring_view messagesKey{ L"messages" };
+static constexpr std::wstring_view choicesKey{ L"choices" };
+static constexpr std::wstring_view roleKey{ L"role" };
+static constexpr std::wstring_view assistantKey{ L"assistant" };
+static constexpr std::wstring_view userKey{ L"user" };
+static constexpr std::wstring_view systemKey{ L"system" };
+
+// endpoints
+static constexpr std::wstring_view githubGraphQLEndpoint{ L"https://api.github.com/graphql" };
+static constexpr std::wstring_view chatCompletionSuffix{ L"/chat/completions" };
+static constexpr std::wstring_view accessTokenEndpoint{ L"https://github.com/login/oauth/access_token" };
+
+// Windows Terminal specific strings
+static constexpr std::wstring_view windowsTerminalUserAgent{ L"Windows Terminal" };
+static constexpr std::wstring_view windowsTerminalIntegrationId{ L"windows-terminal-chat" };
+
+namespace winrt::Microsoft::Terminal::Query::Extension::implementation
+{
+    GithubCopilotBranding::GithubCopilotBranding()
+    {
+        _HeaderIconPath = headerIconPath;
+        _HeaderText = RS_(L"GithubCopilot_HeaderText");
+        _SubheaderText = RS_(L"GithubCopilot_SubheaderText");
+        _BadgeIconPath = badgeIconPath;
+        _ResponseMetaData = RS_(L"GithubCopilot_ResponseMetaData");
+    }
+
+    void GithubCopilotLLMProvider::SetAuthentication(const Windows::Foundation::Collections::ValueSet& authValues)
+    {
+        _httpClient = winrt::Windows::Web::Http::HttpClient{};
+        _httpClient.DefaultRequestHeaders().Accept().TryParseAdd(applicationJsonString);
+        _httpClient.DefaultRequestHeaders().Append(copilotIntegrationIdString, windowsTerminalIntegrationId);
+        _httpClient.DefaultRequestHeaders().UserAgent().TryParseAdd(windowsTerminalUserAgent);
+
+        const auto url = unbox_value_or<hstring>(authValues.TryLookup(urlKey).try_as<IPropertyValue>(), L"");
+        _authToken = unbox_value_or<hstring>(authValues.TryLookup(accessTokenKey).try_as<IPropertyValue>(), L"");
+        _refreshToken = unbox_value_or<hstring>(authValues.TryLookup(refreshTokenKey).try_as<IPropertyValue>(), L"");
+
+        if (!url.empty())
+        {
+            const Windows::Foundation::Uri parsedUrl{ url };
+            const auto randomStateString = unbox_value_or<hstring>(authValues.TryLookup(stateKey).try_as<IPropertyValue>(), L"");
+            // only handle this if the state strings match
+            if (randomStateString == parsedUrl.QueryParsed().GetFirstValueByName(stateKey))
```

Comment:
do we *also* confirm that the random state string matches the one stored on the app?

Reply:
`randomStateString` here _is_ the one stored on the app